### DISCUSSION TITLE
elastic-search index creation

### DISCUSSION THREAD

**Discussion by riyaj8888** [apache/airflow#29294] [General]
### Apache Airflow Provider(s)

elasticsearch

### Versions of Apache Airflow Providers

apache-airflow-providers-elasticsearch == 4.3.3

### Apache Airflow version

2.3.2

### Operating System

windows

### Deployment

Docker-Compose

### What happened

in docker-compose i have added elastic-search as service and using that service i am trying to connect to elastic cluster and create index in DAG file but i am getting following errors
`new connection refused `

`from airflow.providers.elasticsearch.hooks.elasticsearch import ElasticsearchHook

from airflow.operators.dummy_operator import DummyOperator
from datetime import datetime,timedelta
from airflow import DAG
from airflow.operators.python_operator import PythonOperator
from elasticsearch import Elasticsearch

default_args = {"owner":"default",
                "depend_on_past":False,
                "start_date":datetime(2023, 1 ,26),
                "retries":1,
                "retry_delay":timedelta(minutes=15)}
                

mappings = {
    
    "properties": {
        "title": {"type": "text"},
        "description": {"type": "text"},
        "sent-emb": {
            "type": "dense_vector",
            "dims": 768,
            "index": True,
            "similarity": "l2_norm"
        }
}
}

def get_activated_sources():
    dense_vector_dim = 768
    index_name = 'test_es'
    es = Elasticsearch("http://localhost:9200")
    # es.indices.create(index=index_name, body=mappings, ignore=400)
    # es_index(es,dense_vector_dim,index_name)
    # es_hook = ElasticsearchHook(hosts= ['http://localhost:9200'])
    # connection = es_hook.get_conn()
    print(es)

with DAG ("es_dag",default_args=default_args,schedule_interval="@daily",catchup=False) as dag:

    start_task = DummyOperator(task_id="dummpy_task")

    hook_task = PythonOperator(task_id="hook_task",python_callable=get_activated_sources)

    start_task >> hook_task

`

I am able to connect to elastic-search cluster ,but when i am trying run 
es.indices.create(index=index_name, body=mappings, ignore=400) this throwing error.

can we create elastic index inside airflow DAG or not?

thanks 

### What you think should happen instead

same script is able to create elastic index on local machine , but Inside DAG script its throwing error.

### How to reproduce

in docker-compose i have added elastic-search as service and using that service i am trying to connect to elastic cluster and create index in DAG file but i am getting following errors
`new connection refused `

`from airflow.providers.elasticsearch.hooks.elasticsearch import ElasticsearchHook

from airflow.operators.dummy_operator import DummyOperator
from datetime import datetime,timedelta
from airflow import DAG
from airflow.operators.python_operator import PythonOperator
from elasticsearch import Elasticsearch

default_args = {"owner":"default",
                "depend_on_past":False,
                "start_date":datetime(2023, 1 ,26),
                "retries":1,
                "retry_delay":timedelta(minutes=15)}
                

mappings = {
    
    "properties": {
        "title": {"type": "text"},
        "description": {"type": "text"},
        "sent-emb": {
            "type": "dense_vector",
            "dims": 768,
            "index": True,
            "similarity": "l2_norm"
        }
}
}

def get_activated_sources():
    dense_vector_dim = 768
    index_name = 'test_es'
    es = Elasticsearch("http://localhost:9200")
    # es.indices.create(index=index_name, body=mappings, ignore=400)
    # es_index(es,dense_vector_dim,index_name)
    # es_hook = ElasticsearchHook(hosts= ['http://localhost:9200'])
    # connection = es_hook.get_conn()
    print(es)

with DAG ("es_dag",default_args=default_args,schedule_interval="@daily",catchup=False) as dag:

    start_task = DummyOperator(task_id="dummpy_task")

    hook_task = PythonOperator(task_id="hook_task",python_callable=get_activated_sources)

    start_task >> hook_task

`

I am able to connect to elastic-search cluster ,but when i am trying run 
es.indices.create(index=index_name, body=mappings, ignore=400) this throwing error.

can we create elastic index inside airflow DAG or not?

thanks 

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by potiuk** [comment]
This is not an airflow issue - this is at most a troubleshooting ,. that's why I convert it to a discussion. Apparently you have a problem with connecting to the host. If you are looking at connecting to elasticseach to "localhost", then well - this is your problem I guess. I do not know in detail  what you are trying to do, but the "connection refused" error generally indicates that your client tries to connect on given host/port cannot connect. And you have to  make sure that you use the right host/port in your configuration so that your worker can connect to it.

Make sure it is all correctly configured  and your problem will be fixed.

**Comment by potiuk** [comment]
BTW. If you are using docker-compose. Then you should know (and this is a docker question not airflow) that docker on windows run on a separate virtual machine that has a different address and using localhost will connect to the local virtual machine, not to your localhost. Those two are different and have separate ip addresses and networking.

MAYBE then the problem is that you should use the actual IP address of your host machine that you have the service running. But this is basic docker (not airflow) behaviour, that is the same for anything that runs under docker on Windows, so you shoud be aware of it

Example doc about it here: https://www.howtogeek.com/devops/how-to-connect-to-localhost-within-a-docker-container/ ( but you can likely look on your own for other guides on that).

**Reply by riyaj8888** [reply]
thanks for reminding but i am actually doing this way only

**Comment by riyaj8888** [comment]
i am trying to connect localhost using host.docker.internal .
i am able to connect to postgress database as its one of service inside docker-compose. similarly i defined one service for elastic search and trying to connect elastic search. 
`es = Elasticsearch("http://localhost:9200/")`
this connection is successfully , but once i enter 
`es.indices.create(index=index_name, body=mappings, ignore=400)`
getting error and unable to create index.

I have added connection in AIRFLOW UI with following details.
host : elastic(service name in the docker compose) or host.docker.internal
schema: http
port:9200
and others

**Comment by riyaj8888** [comment]
![es-error](https://user-images.githubusercontent.com/29457825/216250330-d35943eb-cf76-4f6c-b40c-8c17d2285f06.PNG)

**Reply by riyaj8888** [reply]
inside host field from AIRFLOW UI  adding host.docker.internal

**Comment by riyaj8888** [comment]
thanks solved the issues. with new version of elastic search
